src/ripple/app/misc/HashRouter.h

```diff
@@ -141,19 +147,19 @@ class HashRouter
 
     int getFlags (uint256 const& key);
 
-    /** Prepare the hash entry to be relayed.
+    /** Determines whether the hashed item should be relayed.
 
         Effects:
 
-            Swaps `peers` with the set of peers that have already
-            been relayed so they can be skipped.
+            If the item should be relayed, this function will not
+            return `true` again until the hold time has expired.
+            The internal set of peers will also be reset.
 
-        @return `true` if the item can be relayed. `false` if not.
-           e.g. Because it has been relayed too recently already.
+        @return A `boost::optional` set of peers which do not need to be
+            relayed to. If the result is uninitialized, the item should
+            _not_ be relayed.
     */
-    bool prepRelay (uint256 const& key, std::set<PeerShortID>& peers);
-
-    bool prepRelay (uint256 const& key);
+    boost::optional<std::set<PeerShortID>> shouldRelay(uint256 const& key);
```

Comment:
I like this new interface. Nice. 

I think we need to document that if the return value is valid then the set of peers that have already been relayed is cleared in this object. It's important to call this out in the doc because with the new interface and name `shouldRelay` looks like a pure function.